src/librustc/traits/project.rs

```diff
@@ -1320,23 +1320,10 @@ fn assoc_ty_def<'cx, 'gcx, 'tcx>(
     let trait_def_id = selcx.tcx().impl_trait_ref(impl_def_id).unwrap().def_id;
     let trait_def = selcx.tcx().trait_def(trait_def_id);
 
-    if !trait_def.is_complete(selcx.tcx()) {
```

Comment:
> It needs to be used in the specialization graph of the same trait.

Yes, that makes sense. And I thought the following example might trigger the condition, but it doesn't `:(`
```rust
#![feature(specialization)]

trait Assoc {
    type Output;
}

impl<T> Assoc for T {
    default type Output = u8;
}

impl Assoc for u32 {
    type Output = u32;
}

impl Assoc for u64 {
    // Shouldn't this cause a reentrant call?
    type Output = <u32 as Assoc>::Output;
}

fn main() {
    let _x: <u32 as Assoc>::Output = 0;
}
```